Counterexample Improvements

ArquintL opened the pull request:
Supersedes #543 with the same feature additions (support for domain values and functions in counterexamples) but with code formatting improvements.
The unit tests could be improved to check for particular counterexamples as soon as [Silver PR # 516](https://github.com/viperproject/silver/pull/516) is merged.

ArquintL commented:
@mschwerhoff I had to create this PR as I cannot push any changes to the existing PR. You might want to look only at the more recent commits as you have already reviewed several commits (look in the old PR)

ArquintL commented:
CI currently fails as it needs the changes from [Silver PR #551](https://github.com/viperproject/silver/pull/551)

mschwerhoff commented on src/main/scala/interfaces/Verification.scala:
```diff
@@ -141,16 +142,18 @@ case class SiliconMappedCounterexample(
   override lazy val toString: String = {
     val buf = converter.modelAtLabel
-      .map(x => s"model at label: ${x._1}\n${x._2.toString}\n")
+      .map(x => s"model at label: ${x._1}\n${ExtractedModel(x._2.entries.map(y => (y._1, interpret(y._2)))).toString}\n")
```
I find `... map { case (foo, bar) => (foo, interpret(bar))}` to be more readable than `y._1 ... y._2` due to the names given to the two tuple components.

mschwerhoff commented on src/main/scala/interfaces/Verification.scala:
```diff
@@ -141,16 +142,18 @@ case class SiliconMappedCounterexample(
+      .map(x => s"model at label: ${x._1}\n${ExtractedModel(x._2.entries.map(y => (y._1, interpret(y._2)))).toString}\n")
       .mkString("\n")
-    s"$buf\non return: \n${converter.extractedModel.toString}"
+    s"$buf\non return: \n${ExtractedModel(converter.extractedModel.entries.map(y => (y._1, interpret(y._2)))).toString}"
```
See previous comment

mschwerhoff commented on src/main/scala/reporting/DomainInterpreter.scala:
```diff
@@ -0,0 +1,51 @@
+  *  @param	T : result type
+  */
+trait AbstractInterpreter[E, I, T] { 
+	def interpret(entry: E, info: I): T
```
Please stick with an indentation of 2 (instead of 4). Probably affects several files.

mschwerhoff commented on src/main/scala/reporting/DomainInterpreter.scala:
```diff
@@ -0,0 +1,51 @@
+				val relevantDomain = domains.filter(x => x.valueName == dom)
+				val domainFunctions = relevantDomain.flatMap(_.functions)
+				val allFunctions = functions ++ domainFunctions
+				val relevantFunctions = allFunctions.filter(x => x.argtypes.length == 1 && x.argtypes.head == sorts.UserSort(Identifier.apply(dom)))
```
I don't see explicit `x.apply(y)` call very often; just `x(y)` appears to be more common.

mschwerhoff commented on src/main/scala/reporting/DomainInterpreter.scala:
```diff
@@ -0,0 +1,51 @@
+				val allFunctions = functions ++ domainFunctions
+				val relevantFunctions = allFunctions.filter(x => x.argtypes.length == 1 && x.argtypes.head == sorts.UserSort(Identifier.apply(dom)))
+				val info = relevantFunctions.map(x => (x,
+					x.apply(Seq(entry)) match {
```
apply?

mschwerhoff commented on src/main/scala/reporting/DomainInterpreter.scala:
```diff
@@ -0,0 +1,51 @@
+					})).toMap
+				ExtendedDomainValueEntry(DomainValueEntry(dom, v), info)
+			case e: VarEntry => interpret(c.extractVal(e), e +: visited)
+			case r: RefEntry => RefEntry(r.name, r.fields.map(x => (x._1, (interpret(x._2._1, r +: visited), x._2._2))))
```
x._1

mschwerhoff commented on src/test/resources/demo-counterexamples/cyclic-ref.vpr:
```diff
@@ -0,0 +1,9 @@
+field next: Ref
```
Shouldn't there be a copyright message at the beginning of each test file? I guess we're not very consistent here, though.

mschwerhoff reviewed (approved):
Looks generally good, but please address the comments before merging the PR.

ArquintL commented:
@mschwerhoff thanks for the review! Your comments are all addressed. In addition, I've tried to add some meaningful tests for the provided examples. You can find the corresponding changeset [here](https://github.com/viperproject/silicon/pull/585/files#diff-2eebf6198d69e46e6a6577ca6d3865c785c305f48afcfeea559721b2b6c52d33). What do you think?

Aurel300 commented:
Maybe this should be rebased since the commits by Cedric were already merged in #528? The last couple of commits from that (Cedric's) PR are actually newer that when this (Fabio's) branch started.

ArquintL commented:
> Maybe this should be rebased since the commits by Cedric were already merged in #528? The last couple of commits from that (Cedric's) PR are actually newer that when this (Fabio's) branch started.

I've already merged in `master` into this PR. I'm not sure why Cedric's commits thus still appear as part of this PR. Do you have an idea how to fix this or should we rather leave it as is and merge this PR?

mschwerhoff reviewed (approved):
Looks good, modulo two copyright headers

ArquintL commented:
@Aurel300 I've now cleaned up the history and rebased it. I've compared the diff before and after rebasing to make sure everything turned out as expected. Could you maybe take a quick look at the changeset to make sure nothing of Cedric's work is dropped by accident?

Aurel300 commented on src/main/scala/reporting/Converter.scala:
```diff
@@ -156,37 +172,72 @@ object Converter {
-      case sorts.Seq(_) => VarEntry(m.toString, s) // will be resolved later
+      case sorts.Seq(_) 
+        | sorts.Set(_)
+        // | sorts.Map(_)
+        // | sorts.Multiset(_)
```
Why are these commented out?

Aurel300 commented on src/main/scala/reporting/Converter.scala:
```diff
@@ -156,37 +172,72 @@ object Converter {
         }
+      case sorts.UserSort(id) => 
+        m match {
+          case ConstantEntry(v) => DomainValueEntry(id.toString, v.split("!").last) //this is a hack for the moment if there is any way to do this differently let me know
```
Is the comment referring to the `.split("!")` bit? Would be better to make this explicit and searchable:
```suggestion
          // TODO: remove this string-based operation
          case ConstantEntry(v) => DomainValueEntry(id.toString, v.split("!").last)
```

Aurel300 commented on src/main/scala/reporting/Converter.scala:
```diff
@@ -156,37 +172,72 @@ object Converter {
+          case ConstantEntry(v) => DomainValueEntry(id.toString, v.split("!").last) //this is a hack for the moment if there is any way to do this differently let me know
+          case _ => OtherEntry(id.toString,"not a constant entry---")
+      }
+      //ISSUE: snap types are translated to domain sorts
```
@mschwerhoff or @ArquintL does the comment here make sense? When can this happen?

Aurel300 commented on src/main/scala/reporting/Converter.scala:
```diff
@@ -269,7 +318,7 @@ object Converter {
             getFunctionValue(model, fname, Seq(entry), to)
           case OtherEntry(t, _) =>
-            OtherEntry(s"SortWrapper($t)", "unapplicable")
+            OtherEntry(s"SortWrapper($t)o", "unapplicable")
```
Is this a typo? Should it be `($to)` instead?

Aurel300 commented on src/main/scala/reporting/Converter.scala:
```diff
@@ -298,9 +344,26 @@ object Converter {
+       
+        try { // many things can go wrong but if they do, we cannot infer anything anyways
+          val recvsort = c.singletonRcvr.get.sort
+          val receivers = (0 to 10).map(x => VarEntry(s"$$Ref!val!$x", recvsort))
```
Is this hardcoded to only look for 10 references?

Aurel300 commented on src/main/scala/reporting/Converter.scala:
```diff
@@ -329,26 +392,67 @@ object Converter {
-      encountered: Set[String]
-  ): ExtractedModelEntry = {
+  
+  def evalPerm(value: Term, model: Model): Option[Rational] = {
```
This looks like a very simple, stripped-down interpreter for counterexample terms. Is there anything in Silicon that could already do this? If not, could this be useful for more than just evaluation permissions?

Aurel300 commented on src/main/scala/reporting/Converter.scala:
```diff
@@ -329,26 +392,67 @@ object Converter {
+      case _: FractionPerm => None
+      case IsValidPermVar(_) => None
+      case IsReadPermVar(_, _) => None
+      case PermTimes(v1, v2) => (evalPerm(v1, model), evalPerm(v2, model)) match {
+        case (Some(x), Some(y)) => Some(x * y)
+        case _ => None
+      }
```
Maybe something like this for all of the operations?

```suggestion
      case PermTimes(v1, v2) =>
        evalPerm(v1, model).map(x => evalPerm(v2, model).map(y => x * y))
```

Aurel300 commented on src/main/scala/reporting/Converter.scala:
```diff
@@ -412,58 +538,53 @@ object Converter {
+    try{
+      Some(symbolConverter.toSort(typ)) // simplify the logic with this
+    }catch{
+      case _: Throwable => None
```
Are we sure this isn't hiding errors we would like to see?

Aurel300 commented:
Do not merge this: @mlimbeck will open another PR based on this one with some fixes.
